### PR TITLE
PRMP-928 Get user login data

### PR DESCRIPTION
**Pull request by steph-torres-nhs**

### CODE REVIEW

File: lambdas/tests/unit/helpers/data/statistic/mock_logs_query_results.py [hunk 1]
```diff
@@ -2,19 +2,24 @@
 USER_ID_2 = "9E7F1235-3DF1-4822-AFFB-C4FCC88C2690"
 HASHED_USER_ID_1 = "3192b6cf7ef953cf1a1f0945a83b55ab2cb8bae95cac6548ae5412aaa4c67677"
 HASHED_USER_ID_2 = "a89d1cb4ac0776e45131c65a69e8b1a48026e9b497c94409e480588418a016e4"
+HASHED_USER_ID_1_WITH_ADMIN_ROLE = HASHED_USER_ID_1 + " - GP_ADMIN - RO76"
+HASHED_USER_ID_1_WITH_PCSE_ROLE = HASHED_USER_ID_1 + " - PCSE - "
+HASHED_USER_ID_2_WITH_CLINICAL_ROLE = HASHED_USER_ID_2 + " - GP_CLINICAL - RO76"
```
Comment (MohammadIqbalAD-NHS) [inline lambdas/tests/unit/helpers/data/statistic/mock_logs_query_results.py:7]: I suggest using f-strings here for improved readability.
```suggestion
HASHED_USER_ID_1_WITH_ADMIN_ROLE = f"{HASHED_USER_ID_1} - GP_ADMIN - RO76"
HASHED_USER_ID_1_WITH_PCSE_ROLE = f"{HASHED_USER_ID_1} - PCSE - "
HASHED_USER_ID_2_WITH_CLINICAL_ROLE = f"{HASHED_USER_ID_2} - GP_CLINICAL - RO76"
```

File: .gitignore [hunk 2]
```diff
@@ -102,6 +102,7 @@ lambdas/tests/unit/helpers/data/pdf/tmp
 /lambdas/package_/
 
 batch_update_progress.json
+data_collection_output.json
```
Comment (NogaNHS) [inline .gitignore:105]: remove

File: lambdas/tests/unit/services/test_bulk_upload_report_service.py [hunk 3]
```diff
@@ -283,12 +283,12 @@ def test_report_handler_with_items_uploads_summary_report_to_bucket(
...
             s3_bucket_name=MOCK_STATISTICS_REPORT_BUCKET_NAME,
-            file_key=f"daily-reports/Bulk upload report for {str(MOCK_END_REPORT_TIME)}.csv",
-            file_name=f"/tmp/Bulk upload report for {str(MOCK_END_REPORT_TIME)}.csv",
+            file_key=f"daily-reports/Bulk upload report for {str(MOCK_START_REPORT_TIME)} to {str(MOCK_END_REPORT_TIME)}.csv",
+            file_name=f"/tmp/Bulk upload report for {str(MOCK_START_REPORT_TIME)} to {str(MOCK_END_REPORT_TIME)}.csv",
         ),
     ]
 
-    bulk_upload_report_service.s3_service.upload_file.has_calls(calls)
+    bulk_upload_report_service.s3_service.upload_file.assert_has_calls(calls)
```
Comment (NogaNHS) [inline lambdas/tests/unit/services/test_bulk_upload_report_service.py:291]: remove, so to no conflict with changes from PRMP-765

File: lambdas/utils/cloudwatch_logs_query.py [hunk 4]
```diff
@@ -43,10 +43,24 @@ class CloudwatchLogsQueryParams:
...
 UniqueActiveUserIds = CloudwatchLogsQueryParams(
-    lambda_name="AuthoriserLambda",
+    lambda_name="TokenRequestHandler",
     query_string="""
-        fields @timestamp, Authorisation.selected_organisation.org_ods_code AS ods_code, Authorisation.nhs_user_id AS user_id
+        fields @timestamp,
+        Authorisation.selected_organisation.org_ods_code AS ods_code,
+        Authorisation.nhs_user_id AS user_id,
+        Authorisation.selected_organisation.role_code AS role_code,
+        Authorisation.repository_role AS role_id
```
Comment (MohammadIqbalAD-NHS) [inline lambdas/utils/cloudwatch_logs_query.py:63]: Is `repository_role` the same as `role_id`? If so, could they both be named identically?

File: lambdas/tests/unit/services/test_bulk_upload_report_service.py (hunk 3 above)
Comment (steph-torres-nhs) [reply lambdas/tests/unit/services/test_bulk_upload_report_service.py:291]: Done

File: lambdas/utils/cloudwatch_logs_query.py (hunk 4 above)
Comment (steph-torres-nhs) [reply lambdas/utils/cloudwatch_logs_query.py:63]: I could remove id, more aligned with user_role than id

**Review by MohammadIqbalAD-NHS — APPROVED** [review]

File: lambdas/utils/cloudwatch_logs_query.py [hunk 5]
```diff
@@ -43,10 +43,24 @@ class CloudwatchLogsQueryParams:
...
+    lambda_name="TokenRequestHandler",
     query_string="""
-        fields @timestamp, Authorisation.selected_organisation.org_ods_code AS ods_code, Authorisation.nhs_user_id AS user_id
+        fields @timestamp,
+        Authorisation.selected_organisation.org_ods_code AS ods_code,
+        Authorisation.nhs_user_id AS user_id,
+        Authorisation.selected_organisation.role_code AS role_code,
+        Authorisation.repository_role AS user_role
         | filter ispresent(ods_code) AND ispresent(user_id)
         | dedup(ods_code, user_id)
```
Comment (bethany-kish-nhs) [inline lambdas/utils/cloudwatch_logs_query.py:65]:
```suggestion
        | dedup(ods_code, user_id, user_role)
```

tested in ndr-test and it was only counting one role per login, Noga suggested trying this and it resolved the issue so multiple roles are counted

File: lambdas/utils/cloudwatch_logs_query.py [hunk 6]
```diff
@@ -43,11 +43,25 @@ class CloudwatchLogsQueryParams:
...
     query_string="""
-        fields @timestamp, Authorisation.selected_organisation.org_ods_code AS ods_code, Authorisation.nhs_user_id AS user_id
+        fields @timestamp,
+        Authorisation.selected_organisation.org_ods_code AS ods_code,
+        Authorisation.nhs_user_id AS user_id,
+        Authorisation.selected_organisation.role_code AS role_code,
+        Authorisation.repository_role AS user_role
         | filter ispresent(ods_code) AND ispresent(user_id)
-        | dedup(ods_code, user_id)
+        | dedup(ods_code, user_id, user_role)
```
Comment (NogaNHS) [inline lambdas/utils/cloudwatch_logs_query.py:65]:
```suggestion
        | dedup(ods_code, user_id, user_role, role_code)
```

**Comment by bethany-kish-nhs** [comment]
Tested on ndr-test with a minor change which has been merged in, can be merged in

**Pull request merged**